src/samplers/mh.jl

```diff
@@ -0,0 +1,243 @@
+  end
+
+  vi = resume_from == nothing ?
+       model() :
+       deepcopy(resume_from.info[:vi])
+
+  if spl.alg.gid == 0
+    runmodel(model, vi, spl)
+  end
+
+  # MH steps
+  if PROGRESS spl.info[:progress] = ProgressMeter.Progress(n, 1, "[MH] Sampling...", 0) end
+  for i = 1:n
+    dprintln(2, "MH stepping...")
+
+    time_elapsed = @elapsed vi = step(model, spl, vi, i == 1)
+    time_total += time_elapsed
+
+    if spl.info[:accept_his][end]     # accepted => store the new predcits
+      samples[i].value = Sample(vi, spl).value
+    else                              # rejected => store the previous predcits
+      samples[i] = samples[i - 1]
+    end
+    samples[i].value[:elapsed] = time_elapsed
+
+    if PROGRESS ProgressMeter.next!(spl.info[:progress]) end
+  end
+
+  println("[MH] Finished with")
+  println("  Running time        = $time_total;")
+  accept_rate = sum(spl.info[:accept_his]) / n  # calculate the accept rate
+  println("  Accept rate         = $accept_rate;")
+
+  setchunksize(default_chunk_size)      # revert global chunk size
+
+  if resume_from != nothing   # concat samples
+    unshift!(samples, resume_from.value2...)
+  end
+  c = Chain(0, samples)       # wrap the result by Chain
+  if save_state               # save state
+    save!(c, spl, model, vi)
+  end
+
+  c
+end
+
+function rand_truncated(dist, lowerbound, upperbound)
+    notvalid = true
+    x = 0.0
+    while (notvalid)
+        x = rand(dist)
+        notvalid = ((x < lowerbound) | (x > upperbound))
+    end
+    return x
+end
+
+assume(spl::Sampler{MH}, dist::Distribution, vn::VarName, vi::VarInfo) = begin
+    if isempty(spl.alg.space) || vn.sym in spl.alg.space
+      vi.index += 1
+      old_val = getval(vi, vn)[1]
+
+      if ~haskey(vi, vn)
+        r = rand(dist)
+        push!(vi, vn, r, dist, spl.alg.gid)
+        spl.info[:proposal_ratio] += (logpdf(dist, old_val) - logpdf(dist, r))
+        spl.info[:cache_updated] = CACHERESET # sanity flag mask for getidcs and getranges
+
+      elseif vn.sym in keys(spl.alg.proposals) # Custom proposal for this parameter
+        proposal = spl.alg.proposals[vn.sym](old_val)
+
+        if typeof(proposal) == Distributions.Normal{Float64} # If Gaussian proposal
+          σ = std(proposal)
+          lb = support(dist).lb
+          ub = support(dist).ub
+          stdG = Normal()
+          r = rand_truncated(proposal, lb, ub)
+          # cf http://fsaad.scripts.mit.edu/randomseed/metropolis-hastings-sampling-with-gaussian-drift-proposal-on-bounded-support/
+          spl.info[:proposal_ratio] += log(cdf(stdG, (ub-old_val)/σ) - cdf(stdG,(lb-old_val)/σ))
+          spl.info[:proposal_ratio] -= log(cdf(stdG, (ub-r)/σ) - cdf(stdG,(lb-r)/σ))
+
+        else # Other than Gaussian proposal
+          r = rand(proposal)
+          if (r < support(dist).lb) | (r > support(dist).ub) # check if value lies in support
+            spl.info[:violating_support] = true
```

Comment:
So you're rejecting out-of-bounds proposals. But don't you need to deal with the acceptance ratio as we did for the Gaussian? (I.e. why is the blog post only relevant to Gaussian proposals)

Reply:
The blog post is relevant for the Gaussian case since we can analytically account for the truncated proposal (which depends on the current value). Maybe this can also be done for other distributions.

Reply:
Yh wouldn't your acceptance ratio be invalid otherwise?

Reply:
The ratio would be valid in the generic proposal case with a Gaussian proposal. But you could potentially be wasting proposals. Let's say we have 10 random variables, if the 10th proposal does not lie on the associated prior's support, we would have to throw all these proposal. With the Gaussian, we can avoid that by sampling from a truncated Gaussian, and correct the acceptance ratio to take into account that this truncation depends on our current value.